### ISSUE TITLE
module federation HMR on host

### ISSUE THREAD

**Issue by nirberko** [nrwl/nx#11089]
## Current Behavior
host app HMR is not working and I need to reload the page by myself. but remote app HMR is working as expected (inside the host app).

## Expected Behavior
when changing the code in the host app, it should reload the app.

## Steps to Reproduce
just look at my configurations:

 for the host `project.json` (I'm using nx server):

```
"serve": {
  "executor": "@nrwl/react:module-federation-dev-server",
  "defaultConfiguration": "development",
  "options": {
    "buildTarget": "client:build",
    "hmr": true,
    "port": 4200
  },
  "configurations": {
    "development": {
      "buildTarget": "client:build:development",
      "open": true,
      "watch": true,
      "liveReload": true
    },
    "production": {
      "buildTarget": "client:build:production",
      "hmr": false
    }
  }
}
```

and this is in the remote `project.json`

```
"serve": {
  "executor": "@nrwl/react:module-federation-dev-server",
  "defaultConfiguration": "development",
  "options": {
    "buildTarget": "assets:build",
    "hmr": true,
    "port": 4201
  },
  "configurations": {
    "development": {
      "buildTarget": "assets:build:development"
    },
    "production": {
      "buildTarget": "assets:build:production",
      "hmr": false
    }
  }
}
```

### Failure Logs
I see this warning in the console. I'm not sure if its related
```
(missing: http://localhost:4200/polyfills.2c951a6de4399979.hot-update.mjs)
    at http://localhost:4200/polyfills.js:21576:25
    at new Promise (<anonymous>)
    at loadUpdateChunk (http://localhost:4200/polyfills.js:21571:19)
    at http://localhost:4200/polyfills.js:22031:28
    at Array.forEach (<anonymous>)
    at __webpack_require__.hmrC.jsonp (http://localhost:4200/polyfills.js:22026:21)
    at http://localhost:4200/polyfills.js:21054:46
    at Array.reduce (<anonymous>)
    at http://localhost:4200/polyfills.js:21050:54
```
### Environment

```
   Node : 17.3.0
   OS   : darwin x64
   npm  : 8.3.0
   
   nx : 14.4.2
   @nrwl/angular : Not Found
   @nrwl/cypress : 14.4.2
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.4.2
   @nrwl/eslint-plugin-nx : 14.4.2
   @nrwl/express : Not Found
   @nrwl/jest : 14.4.2
   @nrwl/js : 14.4.2
   @nrwl/linter : 14.4.2
   @nrwl/nest : Not Found
   @nrwl/next : Not Found
   @nrwl/node : Not Found
   @nrwl/nx-cloud : Not Found
   @nrwl/nx-plugin : Not Found
   @nrwl/react : 14.4.2
   @nrwl/react-native : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : 14.4.2
   @nrwl/web : 14.4.2
   @nrwl/workspace : 14.4.2
   typescript : 4.7.4
   ---------------------------------------
   Community plugins:
```

**Comment by nirberko** [comment]
no solution for this? maybe a workaround with `webpack-merge`? but I seems like its also doesn't working :(

**Comment by nirberko** [comment]
Anyone? is it possible?
next week I'm starting a new position and considering using module federation.
if it's not possible, ill love to know so ill start a regular react project without module federation.

**Comment by ryansevilla** [comment]
I'm seeing the same issue. Would love a fix.

**Comment by g4ngzt4** [comment]
I have the same issue, anyone found a solution?

**Comment by g4ngzt4** [comment]
I managed to fix it by removing the `"polyfills"` from the `build` target for development

**Comment by usyyy** [comment]
> "open": true,
>       "watch": true,
>       "liveReload": true

Could you paste an example?

**Comment by usyyy** [comment]
Has anybody managed to get the remote application to use hmr?

**Comment by jamieatmirza** [comment]
> Has anybody managed to get the remote application(s) to use hmr?

Yes, all our apps now hmr using the [fix/workaround](https://github.com/nrwl/nx/issues/11089#issuecomment-1263495444) above, using either

```bash
npx nx run host:serve --devRemotes=remote-one,remote-two
```

Or by serving an individual remote directly with

```bash
npx nx run remote-one:serve
```

**Comment by ryansevilla** [comment]
Unfortunately this did not work for me. @jamieatmirza was there anything else you had to do to get it work correctly?

**Comment by jamieatmirza** [comment]
> Unfortunately this did not work for me. @jamieatmirza was there anything else you had to do to get it work correctly?

Do you mean the [polyfill workaround](https://github.com/nrwl/nx/issues/11089#issuecomment-1267036922) didn't work for you.

This is the only modification we have made. We are currently on Nx v14.7.9

**Comment by ryansevilla** [comment]
What ended up fixing it for me was also removing the "styles" attribute from the project.json as well. I made two separate build processes, one for local and one for prod, the local one removes polyfills and styles, the prod one keeps them.

**Comment by ghanshyamtanchak** [comment]
> > > I managed to fix it by removing the `"polyfills"` from the `build` target for development
> > 
> > 
> > Could you paste an example? Also, could this introduce any side effects?
> 
> @usyyy I moved the `polyfills` in the `host` apps `project.json` into the `production` configuration. e.g.
> 
> ```json
>     "build": {
>       "executor": "@nrwl/webpack:webpack",
>       "outputs": ["{options.outputPath}"],
>       "defaultConfiguration": "production",
>       "options": {
>         "compiler": "babel",
>         "outputPath": "dist/apps/host",
>         "index": "apps/host/src/index.html",
>         "baseHref": "/",
>         "main": "apps/host/src/main.ts",
>         "tsConfig": "apps/host/tsconfig.app.json",
>         "assets": [
>           "apps/host/src/favicon.ico"
>         ],
>         "styles": [],
>         "scripts": [],
>         "webpackConfig": "apps/host/webpack.config.js",
>         "postcssConfig": "apps/host/postcss.config.js"
>       },
>       "configurations": {
>         "development": {
>           "extractLicenses": false,
>           "optimization": false,
>           "sourceMap": true,
>           "vendorChunk": true
>         },
>         "production": {
>           "polyfills": "apps/host/src/polyfills.ts",
>           "fileReplacements": [
>             {
>               "replace": "apps/host/src/environments/environment.ts",
>               "with": "apps/host/src/environments/environment.prod.ts"
>             }
>           ],
>           "optimization": true,
>           "outputHashing": "all",
>           "sourceMap": false,
>           "namedChunks": false,
>           "extractLicenses": true,
>           "vendorChunk": false,
>           "webpackConfig": "apps/host/webpack.config.prod.js"
>         }
>       }
>     },
> ```

I have tried this solution but it causing when I build with `nx build` @jamieatmirza

**Comment by ghanshyamtanchak** [comment]
I was facing this error but the Production build is working for me to deploy the Nx app
I have follow [https://nx.dev/recipes/module-federation/faster-builds#production-build-and-deployment-with-nx-cloud](url) 

## Add this command to the .yml file

`nx g @nrwl/workspace:run-commands \
deploy \
--project=host \
--command="rm -rf production && mkdir production && cp -r dist/apps/host/* production && cp -r dist/apps/{shop,cart,about} production"`

**_apps/host/webpack.config.prod.js_**

```module.exports = withModuleFederation({
  ...moduleFederationConfig,
  remotes: [
    ['shop', '<DOMAIN_NAME>/shop'],
    ['cart', '<DOMAIN_NAME>/cart'],
    ['about', '<DOMAIN_NAME>/about'],
  ],
});```

**Comment by crescent1994** [comment]
I encountered a similar problem, after enabling HMR, the project can be started normally, but any changes made to any line of code will result in an error: `.mjs` files *must* set `isModule` to `true`.
But when I delete the styles in project.json, it becomes normal

**Comment by JounQin** [comment]
> I encountered a similar problem, after enabling HMR, the project can be started normally, but any changes made to any line of code will result in an error: `.mjs` files _must_ set `isModule` to `true`. But when I delete the styles in project.json, it becomes normal

This should be related to https://github.com/angular/angular-cli/pull/25594/files#diff-e4dd3ccd61cdf3cbbeb04eeccdf65f865c99c1800af922c42fc421cb08f6ccf1R58

**Comment by mattleong** [comment]
Still an issue on `17.0.0-beta.4`. Luckily, this [workaround](https://github.com/nrwl/nx/issues/11089#issuecomment-1267036922) still works.

**Comment by mathis-m** [comment]
HMR for host still not working with `17.3.1`, no errors:
![image](https://github.com/nrwl/nx/assets/11584315/b210d150-6a0c-4fc2-a1bd-2d8257201ff4)

**Comment by johnnycook** [comment]
I have the same problem. I think the rebuild itself works. But the browser just doesn't reload the newly rebuilt resource.

(Using `"@nx/react": "18.0.5"`)

**Comment by ThejanNim** [comment]
Guys, is there Any update for this, I have the same problem and am using Nx version 18.1.2

**Comment by Coly010** [comment]
Can you try this workaround: https://github.com/nrwl/nx/issues/22300#issuecomment-2025228493

The fix mentioned was released in Nx 18.2.0
